posix/unistd.c

```diff
@@ -0,0 +1,612 @@
+    struct oe_utsname uts;
+
+    if ((ret = oe_uname(&uts)) != 0)
+        OE_RAISE_ERRNO(oe_errno);
+
+    oe_strlcpy(name, uts.domainname, len);
+
+    ret = 0;
+
+done:
+    return ret;
+}
+
+static char _cwd[OE_PATH_MAX] = "/";
+static oe_spinlock_t _cwd_lock;
+
+char* oe_getcwd(char* buf, size_t size)
+{
+    char* ret = NULL;
+    char* p = NULL;
+    size_t n;
+    bool locked = false;
+
+    if (buf && size == 0)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    if (!buf)
+    {
+        n = OE_PATH_MAX;
+        p = oe_malloc(n);
+
+        if (!p)
+            OE_RAISE_ERRNO(OE_ENOMEM);
+    }
+    else
+    {
+        n = size;
+        p = buf;
+    }
+
+    oe_conditional_lock(&_cwd_lock, &locked);
+
+    if (oe_strlcpy(p, _cwd, n) >= n)
+        OE_RAISE_ERRNO(OE_ERANGE);
+
+    oe_conditional_unlock(&_cwd_lock, &locked);
+
+    ret = p;
+    p = NULL;
+
+done:
+
+    oe_conditional_unlock(&_cwd_lock, &locked);
+
+    if (p && p != buf)
+        oe_free(p);
+
+    return ret;
+}
+
+int oe_chdir(const char* path)
+{
+    int ret = -1;
+    char real_path[OE_PATH_MAX];
+    struct oe_stat st;
+    bool locked = false;
+
+    /* Resolve to an absolute canonical path. */
+    if (!oe_realpath(path, real_path))
+        return -1;
+
+    /* Fail if unable to stat the path. */
+    if (oe_stat(real_path, &st) != 0)
+    {
+        // oe_errno set by oe_stat().
+        return -1;
+    }
+
+    /* Fail if path not a directory. */
+    if (!OE_S_ISDIR(st.st_mode))
+    {
+        oe_errno = OE_ENOTDIR;
+        return -1;
+    }
+
+    /* Set the _cwd global. */
+    oe_conditional_lock(&_cwd_lock, &locked);
+
+    if (oe_strlcpy(_cwd, real_path, OE_PATH_MAX) >= OE_PATH_MAX)
+        OE_RAISE_ERRNO(OE_ENAMETOOLONG);
+
+    oe_conditional_unlock(&_cwd_lock, &locked);
+
+    ret = 0;
+
+done:
+
+    oe_conditional_unlock(&_cwd_lock, &locked);
+
+    return ret;
+}
+
+unsigned int oe_sleep(unsigned int seconds)
+{
+    const uint64_t ONE_SECOND = 1000;
+    const uint64_t msec = seconds * ONE_SECOND;
+
+    return (oe_sleep_msec(msec) == 0) ? 0 : seconds;
+}
+
+oe_pid_t oe_getpid(void)
+{
+    oe_pid_t ret = 0;
+    oe_posix_getpid(&ret);
+    return ret;
+}
+
+oe_pid_t oe_getppid(void)
+{
+    oe_pid_t ret = 0;
+    oe_posix_getppid(&ret);
+    return ret;
+}
+
+oe_pid_t oe_getpgrp(void)
+{
+    oe_pid_t ret = 0;
+    oe_posix_getpgrp(&ret);
+    return ret;
+}
+
+oe_uid_t oe_getuid(void)
+{
+    oe_uid_t ret = 0;
+    oe_posix_getuid(&ret);
+    return ret;
+}
+
+oe_uid_t oe_geteuid(void)
+{
+    oe_uid_t ret = 0;
+    oe_posix_geteuid(&ret);
+    return ret;
+}
+
+oe_gid_t oe_getgid(void)
+{
+    oe_gid_t ret = 0;
+    oe_posix_getgid(&ret);
+    return ret;
+}
+
+oe_gid_t oe_getegid(void)
+{
+    oe_gid_t ret = 0;
+    oe_posix_getegid(&ret);
+    return ret;
+}
+
+oe_pid_t oe_getpgid(oe_pid_t pid)
+{
+    oe_pid_t ret = -1;
+    oe_pid_t retval = -1;
+
+    if (oe_posix_getpgid(&retval, pid) != OE_OK)
+    {
+        oe_errno = OE_EINVAL;
```

Comment:
>oe_errno = OE_EINVAL; [](start = 8, length = 21)

[POSIX](http://pubs.opengroup.org/onlinepubs/9699919799/functions/getpgid.html) also expects `EPERM` & `ESRCH` errors for `getpgid` but we always condense to `EINVAL`?

Reply:
This checks for the failure of the OCALL itself (where we raise EINVAL). If the OCALL succeeds, it uses errno propagation to get whatever errno was set by the host.